### PR TITLE
Apache Spark tutorial: Part I (data access)

### PR DESCRIPTION
**Pull request by JulienPeloton**
### What this PR brings?

This PR adds a notebook to illustrate the basics of accessing the merged coadd catalogs through Apache Spark. This notebook follows the same steps (and sometimes same documentation) as in the [GCR notebook](https://github.com/LSSTDESC/DC2-analysis/blob/master/tutorials/object_gcr_1_intro.ipynb).

__Learning objectives__:
- Load and efficiently access a DC2 coadd catalog with Apache Spark
- Understand and have references for the catalog schema
- Apply cuts to the catalog using Spark SQL functionalities
- Have an example of quality cuts and simple star/galaxy separation cut
- Distribute the computation and the routine to plot to be faster!

To play with the notebook, you will need a specific DESC+Spark kernel at NERSC. See [desc-spark](https://github.com/LSSTDESC/desc-spark) for its installation (also explained inside the notebook).

### Move to `tutorials`?

I pushed the notebook to the `contributed` folder. I think it would make more sense to have it in `tutorials`. But as far as I understand, the notebooks in `tutorials` are checked by `beavis` using one of the `desc` jupyter kernel. Obviously that won't work for this notebook because it needs a Spark installation (see [desc-spark](https://github.com/LSSTDESC/desc-spark)).

So my question is: in the case it moves to the `tutorials` folder, can we provide a custom DESC+Spark kernel to `beavis`?

### CODE REVIEW

**Comment by JulienPeloton** [comment]
@yymao thanks! I will move it to `tutorials` then, and I can provide the right kernel for this.

**Comment by JulienPeloton** [comment]
Any follow-up on this one?

**Comment by yymao** [comment]
@JulienPeloton Sorry for the delay. When I was about to review this I realized I don't have the `desc-spark` kernel and didn't proceed. 

@heather999 is there a way to put `desc-spark` in the kernel installation script?

**Comment by JulienPeloton** [comment]
@yymao Thanks for the quick feedback! In the meantime you can follow instruction on how to create the kernel (see [here](https://github.com/LSSTDESC/desc-spark#working-at-nersc-jupyterlab) for example). Should be straightforward and automatically appears on the JupyterLab interface, otherwise let me know.

**Comment by JulienPeloton** [comment]
@yymao thanks for the review. Most of the comments has been addressed.
Note that the use of the word "coadd" instead of "object" was taken directly from the GCR notebook (https://github.com/LSSTDESC/DC2-analysis/blob/master/tutorials/object_gcr_1_intro.ipynb).

File: tutorials/object_spark_1_intro.ipynb [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+    "After going through this notebook, you should be able to:\n",
+    "  1. Load and efficiently access a DC2 coadd catalog with Apache Spark\n",
+    "  2. Understand and have references for the catalog schema\n",
+    "  3. Apply cuts to the catalog using Spark SQL functionalities\n",
+    "  4. Have an example of quality cuts and simple star/galaxy separation cut\n",
+    "  5. Distribute the computation and the routine to plot to be faster!\n",
+    "\n",
+    "__Logistics__: This notebook is intended to be run through the JupyterHub NERSC [interface](https://jupyter-dev.nersc.gov). To setup your Spark+DESC environment at NERSC, please follow the instructions available here: [desc-spark](https://github.com/LSSTDESC/desc-spark).\n",
+    "\n",
+    "__Note concerning resources__\n",
```
[the remark this answers is not in the document]
Comment (JulienPeloton) [reply tutorials/object_spark_1_intro.ipynb]: I would leave it as it still reminds the user to take care of the resource :-)
Note that this message is also printed when the kernel is created.

File: tutorials/object_spark_1_intro.ipynb [hunk 2]
```diff
@@ -0,0 +1,680 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": []
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "desc-spark (2.3.2)",
+   "language": "python",
+   "name": "desc-spark"
```
Comment (yymao) [inline tutorials/object_spark_1_intro.ipynb]: when I followed the instruction of spark kernel installation, the kernel installed is called `desc-pyspark`. So this name doesn't match.

Comment (JulienPeloton) [reply tutorials/object_spark_1_intro.ipynb]: OK thanks I will correct both. 

For the kernel name, this raises up a bigger question: Can we deliver a standard kernel for DESC member (as is done for desc-python or desc-stack)? For the moment I am implicitly assuming that people will generate it.

Comment (yymao) [reply tutorials/object_spark_1_intro.ipynb]: I think we should provide `desc-pystack` as a standard desc kernel (cc @heather999 on this).

Note that if one follows the instruction in https://github.com/astrolabsoftware/spark-kernel-nersc, the kernel name will be `desc-pystack`, and hence, regardless whether we provide this kernel as a standard desc kernel, it seems to make sense to use `desc-pystack` in this tutorial notebook.

Comment (JulienPeloton) [reply tutorials/object_spark_1_intro.ipynb]: Yes I agree (though I think you meant `desc-pyspark` for the name). I would be happy to work on providing and delivering such a kernel for the collaboration. Let me open an issue for discussing this.

Comment (yymao) [reply tutorials/object_spark_1_intro.ipynb]: right, sorry for the typo (fixed)

Comment (JulienPeloton) [reply tutorials/object_spark_1_intro.ipynb]: Following #33 , I reran the notebook with the `desc-python-pyspark` kernel.
Information has been updated inside the notebook as well.

**Comment by yymao** [comment]
@JulienPeloton please clear output when committing the notebook. 

And I think the kernel name still doesn't match? (But maybe we should just wait until we settle the kernel name in #33.)

**Comment by JulienPeloton** [comment]
Ah right, I forgot to clean it! Should be done now.
And for the name, yes we should wait for a decision in #33.

**Comment by JulienPeloton** [comment]
@yymao Following #33 and thanks to @heather999, I have re-processed the notebook with the new `desc-pyspark` kernel. To get the kernel installed in your environment, just run on Cori:

```
source /global/common/software/lsst/common/miniconda/kernels/setup.sh
```

File: tutorials/object_spark_1_intro.ipynb [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+    "\n",
+    "# Read the data as DataFrame\n",
+    "df = spark.read.format(\"parquet\").load(datafile)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### DC2 Coadd catalog Schema\n",
```
Comment (yymao) [inline tutorials/object_spark_1_intro.ipynb]: coadd -> object

File: tutorials/object_spark_1_intro.ipynb [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+   "source": [
+    "The DM stack includes functionality to get the tract and patch number corresponding to a certain position `(RA,DEC)`. However, it is out of the scope of this tutorial."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Apache Spark provides `filter` mechanisms, which you can use to speed up data access if you only need a certain chunks of the dataset.\n",
+    "For the coadd catalog, the chunks are broken into `tract` and `patch`, and hence those are the `filters` you can use:"
```
Comment (yymao) [inline tutorials/object_spark_1_intro.ipynb]: coadd -> object

File: tutorials/object_spark_1_intro.ipynb [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+    "For more information about the stellar locus from models and in Run1.1p see the [DC2 Coadd Run1.1p direct access -- color-color stellar locus](DC2%20Coadd%20Run1.1p%20direct%20access%20--%20color-color%20stellar%20locus.ipynb) tutorial\n",
+    "\n",
+    "Q: What else can you do to improve the star selection?"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "__If you want to see more go to [Part II](DC2%20Coadd%20Run1.1p%20GCR%20tutorial%20--%20Part%20II%3A%20Lensing%20Cuts.ipynb)__"
```
Comment (yymao) [inline tutorials/object_spark_1_intro.ipynb]: This part 2 links back to the GCR tutorial -- clarify it here or remove the link.

File: tutorials/object_spark_1_intro.ipynb [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+    "__Learning objectives__:\n",
+    "\n",
+    "After going through this notebook, you should be able to:\n",
+    "  1. Load and efficiently access a DC2 object catalog with Apache Spark\n",
+    "  2. Understand and have references for the catalog schema\n",
+    "  3. Apply cuts to the catalog using Spark SQL functionalities\n",
+    "  4. Have an example of quality cuts and simple star/galaxy separation cut\n",
+    "  5. Distribute the computation and the routine to plot to be faster!\n",
+    "\n",
+    "__Logistics__: This notebook is intended to be run through the JupyterHub NERSC [interface](https://jupyter-dev.nersc.gov) with the `desc-python-spark` kernel. If you want to setup your Spark+DESC environment at NERSC, please follow the instructions available here: [spark-kernel-nersc](https://github.com/astrolabsoftware/spark-kernel-nersc). Note that a general introduction and tutorials for Apache Spark can be found at [astrolabsoftware/spark-tutorials](https://github.com/astrolabsoftware/spark-tutorials) (under construction).\n",
```
Comment (yymao) [inline tutorials/object_spark_1_intro.ipynb]: Update the kernel installation instruction. 
- `desc-python-spark` --> `desc-pyspark`
- User can use the setup script `/global/common/software/lsst/common/miniconda/kernels/setup.sh`

**Comment by JulienPeloton** [comment]
@yymao thanks for spotting all of that. All done.

**Comment by JulienPeloton** [comment]
Thanks @yymao  for all the hard work on reviewing!

**Pull request merged**